client/packages/android/app/src/main/java/org/openmsupply/client/CertWebViewClient.java

```diff
@@ -65,59 +69,105 @@ private X509Certificate get_x509(SslCertificate certificate) {
         }
     }
 
-    public CertWebViewClient(Bridge bridge, File filesDir) {
+    public CertWebViewClient(Bridge bridge, File filesDir, NativeApi nativeApi) {
         super(bridge);
 
+        savedCertFingerprints = bridge.getContext().getSharedPreferences("savedCertFingerprints", Context.MODE_PRIVATE);
+        this.nativeApi = nativeApi;
         this.filesDir = filesDir;
     }
 
-    @SuppressLint("WebViewClientOnReceivedSslError")
-    @Override
-    public void onReceivedSslError(WebView view, SslErrorHandler handler,
-                                   SslError error) {
-        // TODO store and trust previous cert
-        if (!error.getUrl().contains("localhost")) {
-            handler.proceed();
-            return;
-        }
+    private boolean validateLocalCertificate(SslCertificate targetCert) {
         // If there is a ssl error, check if the request was trying to reach our local trusted
         // remote server. For this:
         // 1) load self signed remote server certificate from local storage
         // 2) validate that the request was trying to reach the server by validating that the
         //    target certificate from the request is matching our known self signed certificate
         Certificate selfSignedCert = this.get_self_signed_cert();
-        SslCertificate targetCert = error.getCertificate();
         if (selfSignedCert == null || targetCert == null) {
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
         Certificate targetX509Cert = get_x509(targetCert);
         if (targetX509Cert == null) {
             Log.e(TAG, "Failed to extract x509 request target certificate");
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
 
         try {
             targetX509Cert.verify(selfSignedCert.getPublicKey());
-            handler.proceed();
-            return;
+            return true;
         } catch (Exception e) {
             Log.e(TAG, "Invalid request target certificate" + e);
         }
 
-        super.onReceivedSslError(view, handler, error);
+        return false;
+    }
+
+    private boolean validateNonLocalCertificate(SslCertificate targetCert, NativeApi.omSupplyServer connectedServer) {
+        // Calculate SSL fingerprint
+        PublicKey publicKey = get_x509(targetCert).getPublicKey();
+        MessageDigest md = null;
+        try {
+           md = MessageDigest.getInstance("SHA-256");
+        } catch (Exception e) {
+            Log.e(TAG, "SHA-256 algorithm is missing" + e);
+            return false;
+        }
+        md.update(publicKey.getEncoded());
+        String fingerprint = Base64.encodeToString(md.digest(), Base64.DEFAULT).trim();
```

Comment:
Usually you store a hash in HEX format. However, if I remember correctly Java doesn't have an implementation in the std lib. What I used before is just this method from:
https://stackoverflow.com/questions/9655181/how-to-convert-a-byte-array-to-a-hex-string-in-java